### PR TITLE
[llvm][TableGen] Add a README to the main TableGen folder

### PR DESCRIPTION
**Pull request by DavidSpickett**
Though I doubt that many people will land here directly, I thought it odd that we didn't have one.

The intro I've copied from the programmer's reference and added a simple example. Then some links to resources and tools, which is the main reason to have this page.

### CODE REVIEW

**Comment by DavidSpickett** [comment]
@joker-eph do you know if there's any other MLIR tablegen things worth mentioning? I've just linked to the LSP server for now. Maybe MLIR has a better page than that I should link instead.

**Comment by joker-eph** [comment]
We have our ODS definition docs: https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/  and https://mlir.llvm.org/docs/DefiningDialects/Operations/

This isn't generic "TableGen" but your "Resources for learning the language" aren't all generic either I believe (by the way should this be curated and split into "TableGen" vs "some TableGen backends"?)

**Comment by DavidSpickett** [comment]
Added the MLIR links and split language and backend resources.

TableGen is live in Compiler Explorer now, but I haven't added a "trunk" variant yet. So I'll add links for that in a follow up when that's there.

File: llvm/utils/TableGen/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+The purpose of TableGen is to generate complex output files based on information
+from source files that are significantly easier to code than the output files would be, and also easier to maintain and modify over time.
+
+The information is coded in a declarative style involving classes and records,
+which are then processed by TableGen.
+
+```
+class Hello <string _msg> {
+	string msg = !strconcat("Hello ", _msg);
```
Comment (mshockwave) [inline llvm/utils/TableGen/README.md:11]: nit: why is the indentation here different from that of line 19 and 23?

**Review by mshockwave — APPROVED** [review]
This is pretty neat! LGTM w/ a minor comment.

> TableGen is live in Compiler Explorer now, but I haven't added a "trunk" variant yet. So I'll add links for that in a follow up when that's there.

Please do! CE is easily approachable for most people.

**Review by joker-eph — APPROVED** [review]

File: llvm/utils/TableGen/README.md (hunk 1 above)
Comment (DavidSpickett) [reply llvm/utils/TableGen/README.md]: I wrote it in Compiler Explorer where <tab> is an actual tab, llvm-tblgen outputs 2 space tabs I presume. I've fixed it.

**Pull request merged**